File: packages/flags-react-native/src/evaluator.ts
/**
 * Flag evaluation logic.
 *
 * Mirrors `@signakit/flags-node`'s evaluator so server- and client-side
 * decisions agree for the same userId / attributes / config.
 */

import type {
  ConfigFlag,
  ProjectConfig,
  VariationAllocation,
  RuleType,
  SignaKitDecision,
  SignaKitDecisions,
  UserAttributes,
  VariableValue,
} from './types'
import { hashForTraffic, hashForVariation, hashForDefault } from './hasher'
import { matchesAudiences } from './audience-matcher'
import { BUCKET_SPACE } from './constants'

function findVariationInRanges(allocation: VariationAllocation, bucket: number): string | null {
  for (const range of allocation.ranges) {
    if (bucket >= range.start && bucket <= range.end) {
      return range.variation
    }
  }
  return null
}

export interface EvaluationResult {
  variationKey: string
  enabled: boolean
  ruleKey: string | null
  ruleType: RuleType | null
  variables: Record<string, VariableValue>
}

function resolveVariables(flag: ConfigFlag, variationKey: string): Record<string, VariableValue> {
  if (!flag.variables || flag.variables.length === 0) return {}

  const variation = flag.variations.find((v) => v.key === variationKey)
  const overrides = variation?.variables ?? {}

  const resolved: Record<string, VariableValue> = {}
  for (const def of flag.variables) {
    const override = overrides[def.key]
    resolved[def.key] = override !== undefined ? override : def.defaultValue
  }
  return resolved
}

export function evaluateFlag(
  flag: ConfigFlag,
  userId: string,
  attributes?: UserAttributes
): EvaluationResult | null {
  if (flag.status === 'archived') {
    return null
  }

  if (!flag.running) {
    return {
      variationKey: 'off',
      enabled: false,
      ruleKey: null,
      ruleType: null,
      variables: resolveVariables(flag, 'off'),
    }
  }

  const rules = flag.rules || []
  for (let i = 0; i < rules.length; i++) {
    const rule = rules[i]
    if (!rule) continue

    if (rule.allowlist && rule.allowlist.length > 0) {
      const allowlistEntry = rule.allowlist.find((entry) => entry.userId === userId)
      if (allowlistEntry) {
        return {
          variationKey: allowlistEntry.variation,
          enabled: allowlistEntry.variation !== 'off',
          ruleKey: rule.ruleKey,
          ruleType: rule.ruleType,
          variables: resolveVariables(flag, allowlistEntry.variation),
        }
      }
    }

    if (matchesAudiences(rule.audiences, rule.audienceMatchType, attributes)) {
      const trafficBucket = hashForTraffic(flag.salt, userId)
      const trafficThreshold = Math.floor((rule.trafficPercentage / 100) * BUCKET_SPACE)

      if (trafficBucket < trafficThreshold) {
        const variationBucket = hashForVariation(flag.salt, userId)
        const variation = findVariationInRanges(rule.variationAllocation, variationBucket)

        if (variation) {
          return {
            variationKey: variation,
            enabled: variation !== 'off',
            ruleKey: rule.ruleKey,
            ruleType: rule.ruleType,
            variables: resolveVariables(flag, variation),
          }
        }
      }
    }
  }

  const defaultBucket = hashForDefault(flag.salt, userId)
  const defaultVariation = findVariationInRanges(flag.allocation, defaultBucket)

  if (defaultVariation) {
    return {
      variationKey: defaultVariation,
      enabled: defaultVariation !== 'off',
      ruleKey: null,
      ruleType: null,
      variables: resolveVariables(flag, defaultVariation),
    }
  }

  return {
    variationKey: 'off',
    enabled: false,
    ruleKey: null,
    ruleType: null,
    variables: resolveVariables(flag, 'off'),
  }
}

export function evaluateAllFlags(
  config: ProjectConfig,
  userId: string,
  attributes?: UserAttributes
): SignaKitDecisions {
  const decisions: SignaKitDecisions = {}

  for (const flag of config.flags) {
    const result = evaluateFlag(flag, userId, attributes)
    if (result !== null) {
      const decision: SignaKitDecision = {
        flagKey: flag.key,
        variationKey: result.variationKey,
        enabled: result.enabled,
        ruleKey: result.ruleKey,
        ruleType: result.ruleType,
        variables: result.variables,
      }
      decisions[flag.key] = decision
    }
  }

  return decisions
}

File: packages/flags-golang/internal/evaluator/evaluator.go
// Package evaluator implements the flag evaluation algorithm.
//
// Algorithm (mirrors packages/flags-node/src/evaluator.ts):
//  1. Archived flags return nil (omit from results).
//  2. Non-running flags return a disabled "off" decision.
//  3. For each rule in order:
//     a. Allowlist match by userId wins immediately.
//     b. Audience must match (any/all). If so, check traffic bucket
//        < trafficPercentage/100 * BucketSpace. If in traffic, pick
//        a variation from the rule's allocation ranges.
//  4. If no rule matched, fall back to the flag's default allocation.
//  5. If even that has no matching range, return disabled "off".
package evaluator

import (
	"github.com/signakit/flags-golang/internal/audience"
	"github.com/signakit/flags-golang/internal/hasher"
	"github.com/signakit/flags-golang/internal/types"
)

// EvaluateFlag evaluates a single flag for a user. Returns nil iff the flag is
// archived (the SDK omits archived flags from decideAll).
func EvaluateFlag(flag types.ConfigFlag, userID string, attrs types.UserAttributes) *types.Decision {
	if flag.Status == types.FlagStatusArchived {
		return nil
	}

	if !flag.Running {
		return offDecision(flag, nil, nil)
	}

	for _, rule := range flag.Rules {
		// 3a. Allowlist takes precedence over everything else in the rule.
		if entry := findAllowlistEntry(rule.Allowlist, userID); entry != nil {
			rt := rule.RuleType
			rk := rule.RuleKey
			return &types.Decision{
				FlagKey:      flag.Key,
				VariationKey: entry.Variation,
				Enabled:      entry.Variation != "off",
				RuleKey:      &rk,
				RuleType:     &rt,
				Variables:    resolveVariables(flag, entry.Variation),
			}
		}

		if !audience.MatchesAudiences(rule.Audiences, rule.AudienceMatchType, attrs) {
			continue
		}

		trafficBucket := hasher.HashForTraffic(flag.Salt, userID)
		// trafficPercentage is 0-100; threshold is bucket count.
		threshold := int((rule.TrafficPercentage / 100.0) * float64(hasher.BucketSpace))
		if trafficBucket >= threshold {
			continue
		}

		variationBucket := hasher.HashForVariation(flag.Salt, userID)
		variation, ok := findVariationInRanges(rule.VariationAllocation, variationBucket)
		if !ok {
			continue
		}

		rt := rule.RuleType
		rk := rule.RuleKey
		return &types.Decision{
			FlagKey:      flag.Key,
			VariationKey: variation,
			Enabled:      variation != "off",
			RuleKey:      &rk,
			RuleType:     &rt,
			Variables:    resolveVariables(flag, variation),
		}
	}

	// Default allocation.
	defaultBucket := hasher.HashForDefault(flag.Salt, userID)
	if variation, ok := findVariationInRanges(flag.Allocation, defaultBucket); ok {
		return &types.Decision{
			FlagKey:      flag.Key,
			VariationKey: variation,
			Enabled:      variation != "off",
			RuleKey:      nil,
			RuleType:     nil,
			Variables:    resolveVariables(flag, variation),
		}
	}

	return offDecision(flag, nil, nil)
}

// EvaluateAll evaluates every flag in cfg, omitting archived flags.
func EvaluateAll(cfg *types.ProjectConfig, userID string, attrs types.UserAttributes) types.Decisions {
	out := make(types.Decisions, len(cfg.Flags))
	for _, flag := range cfg.Flags {
		if d := EvaluateFlag(flag, userID, attrs); d != nil {
			out[flag.Key] = *d
		}
	}
	return out
}

func offDecision(flag types.ConfigFlag, ruleKey *string, ruleType *types.RuleType) *types.Decision {
	return &types.Decision{
		FlagKey:      flag.Key,
		VariationKey: "off",
		Enabled:      false,
		RuleKey:      ruleKey,
		RuleType:     ruleType,
		Variables:    resolveVariables(flag, "off"),
	}
}

func findAllowlistEntry(list []types.AllowlistEntry, userID string) *types.AllowlistEntry {
	for i := range list {
		if list[i].UserID == userID {
			return &list[i]
		}
	}
	return nil
}

func findVariationInRanges(alloc types.VariationAllocation, bucket int) (string, bool) {
	for _, r := range alloc.Ranges {
		if bucket >= r.Start && bucket <= r.End {
			return r.Variation, true
		}
	}
	return "", false
}

func resolveVariables(flag types.ConfigFlag, variationKey string) map[string]any {
	if len(flag.Variables) == 0 {
		return map[string]any{}
	}

	var overrides map[string]any
	for _, v := range flag.Variations {
		if v.Key == variationKey {
			overrides = v.Variables
			break
		}
	}

	out := make(map[string]any, len(flag.Variables))
	for _, def := range flag.Variables {
		if overrides != nil {
			if v, ok := overrides[def.Key]; ok {
				out[def.Key] = v
				continue
			}
		}
		out[def.Key] = def.DefaultValue
	}
	return out
}
